Add missing units from DataBrowser

joabakk opened the pull request:
A lot of paths in use are not in the specification, and thus do not have units or other metadata. This PR adds a way to select missing units directly from DataBrowser.

tkurki commented on packages/server-admin-ui/src/views/DataBrowser.js:
```diff
@@ -318,7 +339,37 @@ class DataBrowser extends Component {
+                        onChange={this.handleUnitChange}
+                        >
+                        <option value=''>no unit</option>
+                        <option value='A'>A (Amperes)</option>
```
Just an idea that these could come from the server and could be available for plugins that way also.

joabakk replied:
I agree. I actually found them in an old version of the spec pdf...

joabakk replied:
Aha, https://github.com/SignalK/specification/blob/master/schemas/definitions.json#L87 seems to have them

tkurki commented on src/serverroutes.js:
```diff
@@ -409,6 +410,17 @@ module.exports = function(app, saveSecurityConfig, getSecurityConfig) {
+    res.json(basedeltas)
+  })
+
+  app.put(`${SERVERROUTESPREFIX}/writeBaseDeltasFile`, (req, res, next) => {
```
isnt' this PUTting basedeltas, no need to say _write_ and _file_?

joabakk replied:
agreed and changed

sbender9 commented:
Should we consider just upgrading everyone to use baseDeltas?

We would just have to change the default here

https://github.com/SignalK/signalk-server/blob/4c14d61d9dbdd249ce747d990d9994482e954f07/src/config/config.js#L70

joabakk commented:
I agree @sbender9 we should. I was reluctant to do it here, but it makes everything easier going forward. The old defaults.json should ideally be deleted or renamed afterwards as well, so that the user does not accidentally update values there.

joabakk commented:
This PR sort of depends on #1311

joabakk commented:
I think this is ready for a review now. There should not be too much overhead, but we could consider hiding the feature behind a switch. Also, there are many ways to form the baseDeltas.json file.

tkurki commented:
I think the unit editor takes too much horizontal and vertical space: it makes all rows take up two lines and the unit editor table cell takes up a lot of vertical space

![image](https://user-images.githubusercontent.com/1049678/121821553-6703c280-cca2-11eb-9eab-c08f5952af84.png)

In mobile layout it does not show anything

![image](https://user-images.githubusercontent.com/1049678/121821627-c82b9600-cca2-11eb-8ffa-1906a10061d1.png)

Are you absolutely sure having the popup menus there all the time makes sense? How about small Edit icon button and autosave of the value that the user clicks, so that you don't need the Save button (that can be pretty far away from the value that you changed, possibly hidden below the fold, easy to forget)?

joabakk commented:
I can do a button and autosave. But I don't see how we can save horisontal and vertical space or make the data browser mobile friendly and still descriptive.

joabakk commented:
OK I have amended the dropdown from "rad/s^2: Angular acceleration in radians per second squared" to  "rad/s^2: Angular acceleration", added an "edit units" button and removed the save button, and made it autosave. This is as mobile friendly as I think I can get it. You still risk people confusing deg and rad for angles, but that is the cost.
<img width="1258" alt="Skjermbilde 2021-06-23 kl  22 17 24" src="https://user-images.githubusercontent.com/2553029/123162949-7b4d7980-d471-11eb-8686-6416fbffccfd.png">

tkurki commented:
What's the use case for having degrees there?

joabakk commented:
This is straight from the spec files. Degrees is for latitude and longitude

tkurki commented:
Sure, but that is actually the unit for lat and longitude separately, and not for position. I would leave deg out.

joabakk commented:
Agreed and done

tkurki commented:
If I
- enable edit units
- change some unit
- disable edit units
the units get saved to baseDeltas.json, but the changed unit is not show in DataBrowser.

After server restart
- the unit I added as not show on the list
- after enabled edit units it is show selected in the popup
- changing the unit and disabling edit units saves the unit in baseDeltas.json but does not show it the list
- after enabling edit units the popup shows the previous unit, not the last edited value

tkurki commented:
I created #1328 instead of nitpicking about indentations in this PR - I can wait to merge that for this PR or if you prefer merge that first and then you can rebase this off the formatted master.

joabakk commented:
Thank you @tkurki. I fixed the bugs you found, and I have done the same procedure with success. 

I think the easiest (for me at least) is if we merge this and then merge the prettify. Perhaps even include prettify in the prepublishOnly script?

tkurki commented:
Progress, but I spotted that this sequence does not work
- edit units
- make a change
- reload the page => the change disappears

joabakk commented:
Thank you Teppo. I have changed the behaviour of the edit switch slightly to store the state locally. The behaviour you describe is tricky (for me at least), because the basDelta.js file is updated, but the internal API call for baseDeltas returns 304 even though the source file has changed. I'm not sure how to fix that

tkurki commented:
Can you explain in more detail - what 'internal api call'?

joabakk commented:
Sure. The render function calls '/skServer/basedeltas' to get the user defaults (https://github.com/joabakk/signalk-server-node/blob/add-units/packages/server-admin-ui/src/views/DataBrowser.js#L54). This api call (https://github.com/joabakk/signalk-server-node/blob/add-units/src/serverroutes.js#L414) is what returns 304 even though the baseDeltas.json file has changed

tkurki commented:
The real problem is that /skServer/baseDeltas does not return the updated values even though the baseDeltas.json file has changed. 304 is just an indication of that.

Try in another browser tab or curl - the added/edited units are not there.

joabakk commented:
Ok I think I fixed that. Next issue is that we need to be able to remove the unit

tkurki commented on packages/server-admin-ui/src/views/DataBrowser.js:
```diff
@@ -168,35 +206,150 @@ class DataBrowser extends Component {
+            if (metas.path && metas.value.units) {
+              let units = this.state.unit
+              units[metas.path] = metas.value.units
+              this.setState({...this.state, unit: units})
```
The code above works, but not why you think and it is against React guidelines.
Check this code:
```
x = {a: 1}
y = x
y.a = 2
console.log(x)
//prints { a: 2 }
```
and read
https://stackoverflow.com/questions/37755997/why-cant-i-directly-modify-a-components-state-really/40309023#40309023 and one way to fix it.

tkurki commented:
Now that I have dug into this in more detail I am not quite sure if the approach here quite the right one. Do we want to download all baseDeltas into the admin UI, then add-or-surgically-edit them and send them back to the server 

OR

should the server implement an endpoint, like /skServer/custom/vessels/*/<path>/meta where the admin UI could PATCH metadata ? Then the server could take care of 
- storing the custom unit the way the server wants (now: by updating baseDeltas, tomorrow: maybe something else)
- sending meta deltas about the new unit (not covered in this PR)
- making the new unit immediately available in the http api (not covered in this PR)

This would be more future proof: later changes in how the server manages custom units/metadata would not need any admin ui changes.

tkurki commented:
Stalled, closing.
